### ISSUE TITLE
Nested complex properties are validated even thoug I didn't use `SetValidator`

### ISSUE THREAD

**Issue by marcinjahn** [FluentValidation/FluentValidation.AspNetCore#7]
### FluentValidation.AspNetCore version

11.2.1

### FluentValidation version

11.2.1

### ASP.NET version

.NET 6

### Summary

According to documentation the complex properties nested within the class I am validating will be validated automatically with one of these two configs:

1. Using the (deprecated) config:

   ```cs
   services.AddFluentValidationAutoValidation(config => 
   {
    config.ImplicitlyValidateChildProperties = true;
   });
   ```
2. Using `SetValidator`: `RuleFor(n => n.MyNestedProperty).SetValidator(new MyNestedPropertyValidator());`

I didn't use neither of these configurations, but my complex property is still being validated.

When I put a breakpoint in a vaidator of a nested complex property, it is being executed when I send the request.
Looking at the example in **Steps To Reproduce**, the `SomeDataValidator` is being created, even though I did not configure it.
If I do configure it with `SetValidator`, the class is being created twice.

### Steps to Reproduce

Here's my setup:

```cs
// DTOs
public sealed record MyRequest(string Name, SomeData Data);

public sealed record SomeData(string Data1, int Data2);

// Validators
public class MyRequestValidator : AbstractValidator<MyRequest> 
{
   public MyRequestValidator()
   {
      RuleFor(n => n.Name).NotEmpty();
   }
}

public class SomeDataValidator : AbstractValidator<SomeData> 
{
   public SomeDataValidator()
   {
      RuleFor(n => n.Data1).NotEmpty();
   }
}

// Dependency Injection in Startup.cs
services.AddValidatorsFromAssemblyContaining<MyRequestValidator>();
services.AddFluentValidationAutoValidation(options => 
    { options.DisableDataAnnotationsValidation = true; });

```

**Comment by JeremySkinner** [comment]
You haven't provided enough information to be able to reproduce a problem. Please provide a small, standalone, runnable sample project that reproduces the issue.

Please also provide a full stack trace that shows where the child validator is being called from.

**Comment by marcinjahn** [comment]
> You haven't provided enough information to be able to reproduce a problem. Please provide a small, standalone, runnable sample project that reproduces the issue.
> 
> Please also provide a full stack trace that shows where the child validator is being called from.

Sure, I'll create some example later today.

**Comment by marcinjahn** [comment]
@JeremySkinner, here's the sample that you asked for: https://github.com/marcinjahn/fluent-validation-nested-properties-issue-sample.

Here's the stack dump. It was copied while the execution was at the nested property constructor:

```
new CharacteristicsValidator()at /home/marcin/tmp/fluentnestedtest/Validators.cs:line 8
RuntimeConstructorInfo.Invoke()
CallSiteRuntimeResolver.VisitConstructor()
CallSiteVisitor<RuntimeResolverContext, __Canon>.VisitCallSiteMain()
CallSiteRuntimeResolver.VisitCache()
CallSiteRuntimeResolver.VisitScopeCache()
CallSiteVisitor<RuntimeResolverContext, object>.VisitCallSite()
CallSiteRuntimeResolver.Resolve()
DynamicServiceProviderEngine.<>c__DisplayClass2_0.<RealizeService>b__0()
ServiceProvider.GetService()
ServiceProviderEngineScope.GetService()
ServiceProviderValidatorFactory.CreateInstance()at /Users/jskinner/code/FluentValidation/src/FluentValidation.DependencyInjectionExtensions/ServiceProviderValidatorFactory.cs:line 33
ValidatorFactoryBase.GetValidator()
FluentValidationModelValidator.Validate()
ValidationVisitor.ValidateNode()
ValidationVisitor.VisitComplexType() [2]
ValidationVisitor.VisitImplementation() [2]
ValidationVisitor.Visit() [2]
ValidationVisitor.VisitChildren()
ValidationVisitor.VisitComplexType() [1]
ValidationVisitor.VisitImplementation() [1]
ValidationVisitor.Visit() [1]
ValidationVisitor.Validate()
FluentValidationVisitor.<>n__1()
FluentValidationVisitor.<>c__DisplayClass2_0.<Validate>g__BaseValidate|0()
FluentValidationVisitor.ValidateInternal()
FluentValidationVisitor.Validate()
ObjectModelValidator.Validate()
ParameterBinder.EnforceBindRequiredAndValidate()
ParameterBinder.<BindModelAsync>d__8.MoveNext()
AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.<BindModelAsync>d__8>()
AsyncValueTaskMethodBuilder<ModelBindingResult>.Start<Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.<BindModelAsync>d__8>()
ParameterBinder.BindModelAsync()
ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d>()
AsyncTaskMethodBuilder.Start<Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d>()
ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<CreateBinderDelegate>g__Bind|0()
ControllerActionInvoker.BindArgumentsAsync()
ControllerActionInvoker.Next()
ControllerActionInvoker.InvokeInnerFilterAsync()
ResourceInvoker.Next()
ResourceInvoker.InvokeFilterPipelineAsync()
ResourceInvoker.InvokeAsync()
ControllerRequestDelegateFactory.<>c__DisplayClass10_0.<CreateRequestDelegate>b__0()
EndpointMiddleware.Invoke()
EndpointRoutingMiddleware.Invoke()
DeveloperExceptionPageMiddleware.<Invoke>d__9.MoveNext()
AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__9>()
AsyncTaskMethodBuilder.Start<Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__9>()
DeveloperExceptionPageMiddleware.Invoke()
HostFilteringMiddleware.Invoke()
HostingApplication.ProcessRequestAsync()
HttpProtocol.<ProcessRequests>d__223<HostingApplication.Context>.MoveNext()
AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<HttpProtocol.<ProcessRequests>d__223<HostingApplication.Context>>.ExecutionContextCallback()
ExecutionContext.RunInternal() [4]
AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<HttpProtocol.<ProcessRequests>d__223<HostingApplication.Context>>.MoveNext()
AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<HttpProtocol.<ProcessRequests>d__223<HostingApplication.Context>>.MoveNext()
AwaitTaskContinuation.RunOrScheduleAction() [2]
Task.RunContinuations() [2]
Task.FinishContinuations() [2]
Task<ReadResult>.TrySetResult()
AsyncTaskMethodBuilder<ReadResult>.SetExistingTaskResult()
AsyncValueTaskMethodBuilder<ReadResult>.SetResult()
StreamPipeReader.<<ReadAsync>g__Core|36_0>d.MoveNext()
AsyncTaskMethodBuilder<ReadResult>.AsyncStateMachineBox<StreamPipeReader.<<ReadAsync>g__Core|36_0>d>.ExecutionContextCallback()
ExecutionContext.RunInternal() [3]
AsyncTaskMethodBuilder<ReadResult>.AsyncStateMachineBox<StreamPipeReader.<<ReadAsync>g__Core|36_0>d>.MoveNext()
AsyncTaskMethodBuilder<ReadResult>.AsyncStateMachineBox<StreamPipeReader.<<ReadAsync>g__Core|36_0>d>.MoveNext()
AwaitTaskContinuation.RunOrScheduleAction() [1]
Task.RunContinuations() [1]
Task.FinishContinuations() [1]
Task<int>.TrySetResult()
AsyncTaskMethodBuilder<int>.SetExistingTaskResult()
AsyncValueTaskMethodBuilder<int>.SetResult()
SslStream.<ReadAsyncInternal>d__188<AsyncReadWriteAdapter>.MoveNext()
AsyncTaskMethodBuilder<int>.AsyncStateMachineBox<SslStream.<ReadAsyncInternal>d__188<AsyncReadWriteAdapter>>.ExecutionContextCallback()
ExecutionContext.RunInternal() [2]
AsyncTaskMethodBuilder<int>.AsyncStateMachineBox<SslStream.<ReadAsyncInternal>d__188<AsyncReadWriteAdapter>>.MoveNext()
AsyncTaskMethodBuilder<int>.AsyncStateMachineBox<SslStream.<ReadAsyncInternal>d__188<AsyncReadWriteAdapter>>.MoveNext()
ThreadPool.<>c.<.cctor>b__86_0()
ManualResetValueTaskSourceCore<int>.SignalCompletion()
PoolingAsyncValueTaskMethodBuilder<int>.SetResult()
DuplexPipeStream.<ReadAsyncInternal>d__27.MoveNext()
PoolingAsyncValueTaskMethodBuilder<int>.StateMachineBox<DuplexPipeStream.<ReadAsyncInternal>d__27>.ExecutionContextCallback()
ExecutionContext.RunInternal() [1]
PoolingAsyncValueTaskMethodBuilder<int>.StateMachineBox<DuplexPipeStream.<ReadAsyncInternal>d__27>.MoveNext()
PoolingAsyncValueTaskMethodBuilder<int>.StateMachineBox<DuplexPipeStream.<ReadAsyncInternal>d__27>.System.Threading.IThreadPoolWorkItem.Execute()
ThreadPoolWorkQueue.Dispatch()
PortableThreadPool.WorkerThread.WorkerThreadStart()
Thread.StartCallback()
[Native to Managed Transition]
```

**Comment by JeremySkinner** [comment]
Thanks for the extra details, I'll see if I can find out what's going on.

>  Reading the readme I understood that by default nested complex properties are not being validated.

Yes you're correct - if you're not calling `SetValidator` then I wouldn't expect the nested property to be validated.

**Comment by JeremySkinner** [comment]
Right, I see what's happening - the problem occurs because you're using records rather than regular classes. 

Inside our model validator, we do some checks on the ModelMetadata type, checking whether it's of type `Type` or of type `Property` ([see here](https://github.com/FluentValidation/FluentValidation.AspNetCore/blob/065e2861b91d7d9d4dbb15504767408c98c9686c/src/FluentValidation.AspNetCore/FluentValidationModelValidatorProvider.cs#L176))

However, in the case of records because the child property is defined inline on the record's parameter, the model metadata comes in as type `Parameter` instead, which we currently don't handle. 

TL;DR: FluentValidation.AspNetCore doesn't correctly handle parameters of record type constructors, which is leading to the issue you're seeing. I'll try and get a fix done for this, but the work around for now would be to use a regular class definition instead.

**Comment by JeremySkinner** [comment]
I believe I have a fix in #8, I probably won't have availability to get a new pushed out today, but I'll aim to do so tomorrow.

**Comment by marcinjahn** [comment]
Oh cool, that makes sense. Thanks for such a quick response!

**Comment by JeremySkinner** [comment]
I've pushed out 11.2.2 with this fix
